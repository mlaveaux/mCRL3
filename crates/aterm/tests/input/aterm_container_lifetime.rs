use mcrl3_aterm::Protected;
use mcrl3_aterm::ATermRef;
use mcrl3_aterm::ATerm;
use mcrl3_aterm::Symbol;
use mcrl3_aterm::Term;

fn main() {
    let mut container: Protected<Vec<ATermRef<'static>>> = Protected::new(Vec::new());

    // Add stuff to the container to test soundness
    {
        let mut write = container.write();
        let t = write.protect(&ATerm::constant(&Symbol::new("a", 0)));
        write.push(t);
    }

    // This should not compile
    let t: ATermRef<'static> = {
        let read = container.read();
        read[0].copy()
    };
    println!("Term: {:?}", t);

    // This should not compile either
    let t: ATermRef<'static> = {
        let mut write = container.write();
        write.pop().unwrap()
    };
    println!("Term: {:?}", t);
}