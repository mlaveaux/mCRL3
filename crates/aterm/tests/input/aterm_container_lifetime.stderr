error[E0597]: `container` does not live long enough
  --> tests/input/aterm_container_lifetime.rs:19:20
   |
8  |     let mut container: Protected<Vec<ATermRef<'static>>> = Protected::new(Vec::new());
   |         ------------- binding `container` declared here
...
18 |     let t: ATermRef<'static> = {
   |            ----------------- type annotation requires that `container` is borrowed for `'static`
19 |         let read = container.read();
   |                    ^^^^^^^^^ borrowed value does not live long enough
...
30 | }
   | - `container` dropped here while still borrowed

error[E0597]: `container` does not live long enough
  --> tests/input/aterm_container_lifetime.rs:26:25
   |
8  |     let mut container: Protected<Vec<ATermRef<'static>>> = Protected::new(Vec::new());
   |         ------------- binding `container` declared here
...
25 |     let t: ATermRef<'static> = {
   |            ----------------- type annotation requires that `container` is borrowed for `'static`
26 |         let mut write = container.write();
   |                         ^^^^^^^^^ borrowed value does not live long enough
...
30 | }
   | - `container` dropped here while still borrowed

error[E0502]: cannot borrow `container` as mutable because it is also borrowed as immutable
  --> tests/input/aterm_container_lifetime.rs:26:25
   |
18 |     let t: ATermRef<'static> = {
   |            ----------------- type annotation requires that `container` is borrowed for `'static`
19 |         let read = container.read();
   |                    --------- immutable borrow occurs here
...
26 |         let mut write = container.write();
   |                         ^^^^^^^^^^^^^^^^^ mutable borrow occurs here
